README
This is version of 3.2.x of at, for running commands at a
specified time.  To install, do a

$ ./configure

You might want to change the default maximum load at which batch
jobs are still started by specifying --with-loadavg_mx=... as
argument to configure.  The default is 0.8, so that, normally, no
batch job will be started when there's still activity going on.
For an SMP system, you will want to increase this.

Then, do

$ make

and, as root,

# make install

at jobs are run by the atd daemon, which is normally started at boot
time.

Bug reports to:
	http://bugs.debian.org/ (Debian bug tracking system)
	or
	at@packages.debian.org (Debian at package maintainers)

If at all possible, I'd appreciate you telling me which version you
found a bug in; run at -V to find out which one.

The Homepage and place for news is at:
        http://blog.calhariz.com/tag/at

The latest source code and old relases are available at the git repository:
        https://salsa.debian.org/debian/at.git/

If you need to contact the developer of the software please use:
        jose (at) calhariz.com

If you try to use at(1) on a system which does not support setreuid(2),
i.e. if you get an error whch looks vaguely like

undefined symbol _setreuid referenced from text segment

DO NOT try to install at on your system by removing all that strange
PRIV stuff.  You will install a rather glaring security hole that way.


ChangeLog
at 3.0:
	at now has a lex/yacc grammer, which is an extension of POSIX.2 to
	something more sensible.  There's an atd, which can still simulate
	atrun with the -s option.

at 3.1:
	Tried to fix some races in the atd code by renaming
	running jobs to "=" (thanks to aj@arthur.pfalz.de for
	pointing out the bug).

at 3.1.1:
	The last fix didn't work.  New locking scheme: as soon as
	a job starts, the file a00beef is hardlinked to =00beef.
	a00beef is unlinked as soon as actual shell execution
	starts.  "=" - files are deleted as soon as their job
	ends, or an hour after they should have run.

at 3.1.2:
	cleanups:  indent the source with OTBS
	remove output file if it already exists, with a warning.
	log attempts to run job twice
	added -b option to atd
	removed -b option from at (now batch calls "at -qb now")
	revamp run_loop so -b works correctly

at 3.1.3:
	Fixed "at 07:30" functionality for implicit tomorrow
	Include yywrap if none can be found

at 3.1.4:
	Set default batch maximum to 0.8
	Changed install group to 'root'

at 3.1.5:
	Merged Bjorn Helgaas' (helgaas@tomahawk.rsn.hp.com) patches:
		* Use setresuid(2) if setreuid(2) is not available
		  (HP-UX 10)
		* Use "chown user:group" instead of "chown user.group" for
		  HP-UX compatibility
		* RELINQUISH_PRIVS slightly earlier in at.c,
		  since it doesn't depend on
		* Change "volatile static" to "static volatile"
		* Test for HAVE_ATTRIBUTE_NORETURN instead of __GNUC__
		  in daemon.h
	Don't export SHELLOPTS (read-only variable in bash)

at 3.1.6:
	Made 'noon', 'teatime' and 'midnight' work.
	Reran autoheader so HP-UX compilation works.

at 3.1.7:
	Made parser case-independent by specifying -i to flex

at 3.1.8:
	Fixed spelling mistake in manpage
	Created at_allow(5) and at_deny(5) manpages
	Fixed bug where 'next monday' could be a synonym for today
	Fixed spelling bug in parser (misspelled February)
	Removed obsolete '-v' flag from atq

at 3.1.11:

  Ansgar Burchardt (15):
      correct comment in atd.c
      rc: do not set -e
      panic.c: switch permissions before unlink
      update configure script
      update config.guess and config.sub
      Makefile.in: do not remove debian/tmp
      at.c: add -t option
      document new -t option
      remove unused function parsetimespec
      at.c: exit successfully when -V is given
      fix some typos
      update contact address
      update copyright information
      set version to 3.1.11
      remove $Id$ tags

  Cyril Brulebois (7):
      Make Copyright a regular file.
      Copy debian/rc to rc, it's not Debian-specific.
      Copy debian/pam to pam.conf, it's not Debian-specific.
      Remove the debian/ directory.
      Fix LFILE/ATJOB_DIR usage in nextjob().
      Fix missing panic.h include.
      Remove unused "static const char *svnid"

at 3.1.12:

  Ansgar Burchardt (18):
      do not increment year twice
      do not increment day of month twice
      update documentation on removed environment variables
      add -f option to stay in foreground
      error when day of month is 0
      make less assumptations on open file descriptors
      use common-session-noninteractive
      make parsetest buildable
      pass -i option to sendmail
      use pid file in init script
      add status action to init script
      make stdout, stderr available to sendmail process
      default queues are no longer configurable
      document -b option
      pass current time as a parameter to parsetime()
      document 4-digit years and ISO date format
      add test suite
      update copyright information

at 3.1.13:

  Ansgar Burchardt (18):
        parsetime.pl: add test case
        correct error message when initgroups fails
        change owner of output file
        do not open output file twice
        use strdup()
        at.1.in: Add references to at.allow(5), at.deny(5)
        Sort references in SEE ALSO section.
        man page for atrun is in section 8
        at.1.in: Document -M option.
        at.1.in: Remove a dot from mid-sentence.
        Rename at_allow.5.in to at.allow.5.in
        Remove useless chdir()
        Create hardlink as priviledged user
        Update my email address.
        atd.c: variable "rc" is always used
        No longer mention atrun in documentation.
        do not abort when job is removed just before we run it
        do not strip binaries on install

  Mike Miller (1):
        Add support for -r option required by POSIX.

  Reuben Thomas (2):
        De-duplicate and improve documentation of /etc/at.{deny,allow}.
        Clarify an error message.

at 3.1.14:

  Ansgar Burchardt (5):
        Add systemd service file.
        Initialize jobno before using it.
        No longer try to compile and run a trivial C program.
        Add -M option to usage message.
        Use LDFLAGS when linking programs.

at 3.1.15 (2014-08-10):

  Ansgar Burchardt (3):
        pam.conf: require pam_loginuid.so
        atd.service.in: Add Documentation field.
        atd.service.in: Set IgnoreSIGPIPE=false

at 3.1.16 (2014-09-30):

  Ansgar Burchardt (1):
        at: only retain variables whose name consists of alphanumerics and underscores

at 3.1.17 (2015-08-31):

  Jose M Calhariz:
	at.1.in: Document behaviour when using past time, thanks Kelly Price (Closes: #639900).
	Makefile.in: On distclean remove at.allow.5, auto generated file.

at 3.1.18 (2015-12-06)

  Jose M Calhariz:
	Add support for SELinux, multiples files

at 3.1.19 (2016-03-20)
  Jose M Calhariz:
	Fix configure handling of flag --without-selinux.

(...)

at 3.1.21 (2018-07-23):
  Jose M Calhariz:
    832368-Using_of_the_meaningless_fcntl
	Fix call of fcntl by replacing (long) 1 with FD_CLOEXEC

    892819-at__improvements_to_atd.service
	Improve atd.service, see bug report 892819

    885891-at__stale_batchjobs_after_reboot
	Remove stale at jobs after a boot.

    897669-897670-Some_fixes_in_the_manuals
	Fix some warnings in manpages at.1 and atd.8


panic.c
/* 
 *  panic.c - terminate fast in case of error
 *  Copyright (C) 1993  Thomas Koenig
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

/* System Headers */

#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif

#include <stdio.h>
#include <stdlib.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAVE_STDARG_H
#include <stdarg.h>
#endif

/* Local headers */

#include "panic.h"
#include "privs.h"
#include "at.h"

/* File scope variables */

/* External variables */

/* Global functions */

void
panic(char *a)
{
/* Something fatal has happened, print error message and exit.
 */
    fprintf(stderr, "%s: %s\n", namep, a);
    if (fcreated) {
	setregid(real_gid, effective_gid);
	unlink(atfile);
	setregid(effective_gid, real_gid);
    }

    exit(EXIT_FAILURE);
}

void
perr(const char *fmt,...)
{
/* Some operating system error; print error message and exit.
 */
    char buf[1024];
    va_list args;

    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);

    perror(buf);
    if (fcreated) {
        setregid(real_gid, effective_gid);
	unlink(atfile);
        setregid(effective_gid, real_gid);
    }

    exit(EXIT_FAILURE);
}

void
usage(void)
{
/* Print usage and exit.
 */
    fprintf(stderr, "Usage: at [-V] [-q x] [-f file] [-u username] [-mMlbv] timespec ...\n"
            "       at [-V] [-q x] [-f file] [-u username] [-mMlbv] -t time\n"
    	    "       at -c job ...\n"
	    "       at [-V] -l [-o timeformat] [job ...]\n"
	    "       atq [-V] [-q x] [-o timeformat] [job ...]\n"
	    "       at [ -rd ] job ...\n"
	    "       atrm [-V] job ...\n"
	    "       batch\n");
    exit(EXIT_FAILURE);
}


daemon.c
/* 
 *  daemon.c : Handle daemon initialization
 *  Copyright (C) 1996 Thomas Koenig
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, 
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>

#ifdef HAVE_STDARG_H
#include <stdarg.h>
#endif

#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <fcntl.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAVE_SYSLOG_H
#include <syslog.h>
#endif

#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif

#include "daemon.h"
#include "privs.h"

int daemon_debug = 0;
int daemon_foreground = 0;

static int
lock_fd(int fd)
{
    struct flock lock;

    lock.l_type = F_WRLCK;
    lock.l_whence = SEEK_SET;
    lock.l_start = 0;
    lock.l_len = 0;

    return fcntl(fd, F_SETLK, &lock);
}

void
perr(const char *fmt,...)
{
    char buf[1024];
    va_list args;

    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);

    if (daemon_debug) {
	perror(buf);
    } else
	syslog(LOG_ERR, "%s: %m", buf);

    exit(EXIT_FAILURE);
}

void
lerr(const char *fmt,...)
{
    char buf[1024];
    va_list args;

    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);

    if (daemon_debug) {
	perror(buf);
    } else
	syslog(LOG_ERR, "%s: %m", buf);
}

void
pabort(const char *fmt,...)
{
    char buf[1024];
    va_list args;

    va_start(args, fmt);
    vsnprintf(buf, sizeof(buf), fmt, args);
    va_end(args);

    if (daemon_debug) {
	fprintf(stderr, "%s\n", buf);
    } else
	syslog(LOG_ERR, "%s", buf);

    exit(EXIT_FAILURE);
}

void
daemon_setup()
{
    /* Set up standard daemon environment */
    pid_t pid;
    mode_t old_umask;
    int fd, devnull;
    FILE *fp;

    if (!daemon_debug) {
	devnull = open("/dev/null", O_RDWR);
	if (devnull == -1) {
	    perr("Error redirecting I/O");
	}

	if ((dup2(devnull, 0) == -1) ||
	    (dup2(devnull, 1) == -1) ||
	    (dup2(devnull, 2) == -1)) {
	    close(devnull);
	    perr("Error redirecting I/O");
	} else {
	    close(devnull);
	}
    }

    if (daemon_foreground)
	pid = getpid();
    else {
	pid = fork();
	if (pid == -1) {
	    perr("Cannot fork");
	} else if (pid != 0) {
	    exit(0);
	}
	(void) setsid();
    }
    old_umask = umask(S_IWGRP | S_IWOTH);

    PRIV_START

    fd = open(PIDFILE, O_RDWR | O_CREAT | O_EXCL, S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);

    if (fd == -1) {

	if (errno != EEXIST)
	    perr("Cannot open " PIDFILE);

	if ((fd = open(PIDFILE, O_RDWR)) < 0)
	    perr("Cannot open " PIDFILE);

	fp = fdopen(fd, "rw");
	if (fp == NULL) {
	    perr("Cannot open " PIDFILE " for reading");
	}
	pid = -1;
	if ((fscanf(fp, "%d", &pid) != 1) || (pid == getpid())
	    || (lock_fd(fileno(fp)) == 0)) {
	    int rc;

	    syslog(LOG_NOTICE, "Removing stale lockfile for pid %d", pid);

	    rc = unlink(PIDFILE);

	    if (rc == -1) {
		perr("Cannot unlink " PIDFILE);
	    }
	} else {
	    pabort("Another atd already running with pid %d", pid);
	}
	fclose(fp);

	unlink(PIDFILE);
	fd = open(PIDFILE, O_RDWR | O_CREAT | O_EXCL,
		  S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH);


	if (fd == -1)
	    perr("Cannot open " PIDFILE " the second time round");

    }

    if (lock_fd(fd) == -1)
	perr("Cannot lock " PIDFILE);

    fp = fdopen(fd, "w");
    if (fp == NULL)
	perr("Special weirdness: fdopen failed");

    fprintf(fp, "%d\n", getpid());

    /* We do NOT close fd, since we want to keep the lock. However, we don't
     * want to keep the file descriptor in case of an exec().
     */
    fflush(fp);
    fcntl(fd, F_SETFD, FD_CLOEXEC);
    PRIV_END

    /* See the above comment. */
    /* coverity[leaked_storage: FALSE] */
    return;
}

void 
daemon_cleanup()
{
    PRIV_START

	unlink(PIDFILE);

    PRIV_END
}


posixtm.c
/*
 * This file is taken from coreutils 7.4
 */

/* Parse dates for touch and date.

   Copyright (C) 1989, 1990, 1991, 1998, 2000, 2001, 2002, 2003, 2004,
   2005, 2006, 2007 Free Software Foundation Inc.

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

/* Yacc-based version written by Jim Kingdon and David MacKenzie.
   Rewritten by Jim Meyering.  */

#include "config.h

#include "posixtm.h"

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <string.h>

#if USE_UNLOCKED_IO
# include "unlocked-io.h"
#endif

/* ISDIGIT differs from isdigit, as follows:
   - Its arg may be any int or unsigned int; it need not be an unsigned char
     or EOF.
   - It's typically faster.
   POSIX says that only '0' through '9' are digits.  Prefer ISDIGIT to
   isdigit unless it's important to use the locale's definition
   of `digit' even when the host does not conform to POSIX.  */
#define ISDIGIT(c) ((unsigned int) (c) - '0' <= 9)

/*
  POSIX requires:

  touch -t [[CC]YY]mmddhhmm[.ss] FILE...
    8, 10, or 12 digits, followed by optional .ss
    (PDS_LEADING_YEAR | PDS_CENTURY | PDS_SECONDS)

  touch mmddhhmm[YY] FILE... (obsoleted by POSIX 1003.1-2001)
    8 or 10 digits, YY (if present) must be in the range 69-99
    (PDS_TRAILING_YEAR | PDS_PRE_2000)

  date mmddhhmm[[CC]YY]
    8, 10, or 12 digits
    (PDS_TRAILING_YEAR | PDS_CENTURY)

*/

static int
year (struct tm *tm, const int *digit_pair, size_t n, unsigned int syntax_bits)
{
  switch (n)
    {
    case 1:
      tm->tm_year = *digit_pair;
      /* Deduce the century based on the year.
	 POSIX requires that 00-68 be interpreted as 2000-2068,
	 and that 69-99 be interpreted as 1969-1999.  */
      if (digit_pair[0] <= 68)
	{
	  if (syntax_bits & PDS_PRE_2000)
	    return 1;
	  tm->tm_year += 100;
	}
      break;

    case 2:
      if (! (syntax_bits & PDS_CENTURY))
	return 1;
      tm->tm_year = digit_pair[0] * 100 + digit_pair[1] - 1900;
      break;

    case 0:
      {
	time_t now;
	struct tm *tmp;

	/* Use current year.  */
	time (&now);
	tmp = localtime (&now);
	if (! tmp)
	  return 1;
	tm->tm_year = tmp->tm_year;
      }
      break;

    default:
      abort ();
    }

  return 0;
}

static int
posix_time_parse (struct tm *tm, const char *s, unsigned int syntax_bits)
{
  const char *dot = NULL;
  int pair[6];
  int *p;
  size_t i;

  size_t s_len = strlen (s);
  size_t len = (((syntax_bits & PDS_SECONDS) && (dot = strchr (s, '.')))
		? (size_t) (dot - s)
		: s_len);

  if (len != 8 && len != 10 && len != 12)
    return 1;

  if (dot)
    {
      if (!(syntax_bits & PDS_SECONDS))
	return 1;

      if (s_len - len != 3)
	return 1;
    }

  for (i = 0; i < len; i++)
    if (!ISDIGIT (s[i]))
      return 1;

  len /= 2;
  for (i = 0; i < len; i++)
    pair[i] = 10 * (s[2*i] - '0') + s[2*i + 1] - '0';

  p = pair;
  if (syntax_bits & PDS_LEADING_YEAR)
    {
      if (year (tm, p, len - 4, syntax_bits))
	return 1;
      p += len - 4;
      len = 4;
    }

  /* Handle 8 digits worth of `MMDDhhmm'.  */
  tm->tm_mon = *p++ - 1;
  tm->tm_mday = *p++;
  tm->tm_hour = *p++;
  tm->tm_min = *p++;
  len -= 4;

  /* Handle any trailing year.  */
  if (syntax_bits & PDS_TRAILING_YEAR)
    {
      if (year (tm, p, len, syntax_bits))
	return 1;
    }

  /* Handle seconds.  */
  if (!dot)
    {
      tm->tm_sec = 0;
    }
  else
    {
      int seconds;

      ++dot;
      if (!ISDIGIT (dot[0]) || !ISDIGIT (dot[1]))
	return 1;
      seconds = 10 * (dot[0] - '0') + dot[1] - '0';

      tm->tm_sec = seconds;
    }

  return 0;
}

/* Parse a POSIX-style date, returning true if successful.  */

bool
posixtime (time_t *p, const char *s, unsigned int syntax_bits)
{
  struct tm tm0;
  struct tm tm1;
  struct tm const *tm;
  time_t t;

  if (posix_time_parse (&tm0, s, syntax_bits))
    return false;

  tm1 = tm0;
  tm1.tm_isdst = -1;
  t = mktime (&tm1);

  if (t != (time_t) -1)
    tm = &tm1;
  else
    {
      /* mktime returns -1 for errors, but -1 is also a valid time_t
	 value.  Check whether an error really occurred.  */
      tm = localtime (&t);
      if (! tm)
	return false;
    }

  /* Reject dates like "September 31" and times like "25:61".  */
  if ((tm0.tm_year ^ tm->tm_year)
      | (tm0.tm_mon ^ tm->tm_mon)
      | (tm0.tm_mday ^ tm->tm_mday)
      | (tm0.tm_hour ^ tm->tm_hour)
      | (tm0.tm_min ^ tm->tm_min)
      | (tm0.tm_sec ^ tm->tm_sec))
    return false;

  *p = t;
  return true;
}

#ifdef TEST_POSIXTIME
/*
    Test mainly with syntax_bits == 13
    (aka: (PDS_LEADING_YEAR | PDS_CENTURY | PDS_SECONDS))

    This test data assumes Universal Time, e.g., TZ="UTC0".

    This test data also assumes that time_t is signed and is at least
    39 bits wide, so that it can represent all years from 0000 through
    9999.  A host with 32-bit signed time_t can represent only time
    stamps in the range 1901-12-13 20:45:52 through 2038-01-18
    03:14:07 UTC, assuming POSIX time_t with no leap seconds, so test
    cases outside this range will not work on such a host.

    Also, the first two lines of test data assume that the current
    year is 2002.

BEGIN-DATA
12131415.16     13   1039788916 Fri Dec 13 14:15:16 2002
12131415.16     13   1039788916 Fri Dec 13 14:15:16 2002
000001010000.00 13 -62167132800 Sun Jan  1 00:00:00 0000
190112132045.52 13  -2147483648 Fri Dec 13 20:45:52 1901
190112132045.53 13  -2147483647 Fri Dec 13 20:45:53 1901
190112132046.52 13  -2147483588 Fri Dec 13 20:46:52 1901
190112132145.52 13  -2147480048 Fri Dec 13 21:45:52 1901
190112142045.52 13  -2147397248 Sat Dec 14 20:45:52 1901
190201132045.52 13  -2144805248 Mon Jan 13 20:45:52 1902
196912312359.59 13           -1 Wed Dec 31 23:59:59 1969
197001010000.00 13            0 Thu Jan  1 00:00:00 1970
197001010000.01 13            1 Thu Jan  1 00:00:01 1970
197001010001.00 13           60 Thu Jan  1 00:01:00 1970
197001010100.00 13         3600 Thu Jan  1 01:00:00 1970
197001020000.00 13        86400 Fri Jan  2 00:00:00 1970
197002010000.00 13      2678400 Sun Feb  1 00:00:00 1970
197101010000.00 13     31536000 Fri Jan  1 00:00:00 1971
197001000000.00 13            * *
197000010000.00 13            * *
197001010000.60 13            * *
197001010060.00 13            * *
197001012400.00 13            * *
197001320000.00 13            * *
197013010000.00 13            * *
203801190314.06 13   2147483646 Tue Jan 19 03:14:06 2038
203801190314.07 13   2147483647 Tue Jan 19 03:14:07 2038
203801190314.08 13   2147483648 Tue Jan 19 03:14:08 2038
999912312359.59 13 253402300799 Fri Dec 31 23:59:59 9999
1112131415      13   1323785700 Tue Dec 13 14:15:00 2011
1112131415.16   13   1323785716 Tue Dec 13 14:15:16 2011
201112131415.16 13   1323785716 Tue Dec 13 14:15:16 2011
191112131415.16 13  -1831974284 Wed Dec 13 14:15:16 1911
203712131415.16 13   2144326516 Sun Dec 13 14:15:16 2037
3712131415.16   13   2144326516 Sun Dec 13 14:15:16 2037
6812131415.16   13   3122633716 Thu Dec 13 14:15:16 2068
6912131415.16   13     -1590284 Sat Dec 13 14:15:16 1969
7012131415.16   13     29945716 Sun Dec 13 14:15:16 1970
1213141599       2    945094500 Mon Dec 13 14:15:00 1999
1213141500       2    976716900 Wed Dec 13 14:15:00 2000
END-DATA

*/

# define MAX_BUFF_LEN 1024

int
main (void)
{
  char buff[MAX_BUFF_LEN + 1];

  buff[MAX_BUFF_LEN] = 0;
  while (fgets (buff, MAX_BUFF_LEN, stdin) && buff[0])
    {
      char time_str[MAX_BUFF_LEN];
      unsigned int syntax_bits;
      time_t t;
      if (sscanf (buff, "%s %u", time_str, &syntax_bits) != 2)
	printf ("*\n");
      else
	{
	  printf ("%-15s %2u ", time_str, syntax_bits);
	  if (posixtime (&t, time_str, syntax_bits))
	    printf ("%12ld %s", (long int) t, ctime (&t));
	  else
	    printf ("%12s %s", "*", "*\n");
	}
    }
  exit (0);

}
#endif

/*
Local Variables:
compile-command: "gcc -DTEST_POSIXTIME -g -O -Wall -W posixtm.c"
End:
*/
